cip36/src/lib.rs

```diff
@@ -0,0 +1,602 @@
+    }
+}
+
+impl std::convert::TryFrom<&Metadata> for DeregistrationCbor {
+    type Error = DeserializeError;
+
+    fn try_from(metadata: &Metadata) -> Result<Self, Self::Error> {
+        use cardano_multiplatform_lib_core::error::Key;
+        let dereg_metadatum = metadata
+            .get(&KEY_DEREGISTRATION_LABEL)
+            .ok_or_else(|| DeserializeFailure::MandatoryFieldMissing(Key::Uint(KEY_DEREGISTRATION_LABEL)))?;
+        let witness_metadatum = metadata
+            .get(&DEREGISTRATION_WITNESS_LABEL)
+            .ok_or_else(|| DeserializeFailure::MandatoryFieldMissing(Key::Uint(DEREGISTRATION_WITNESS_LABEL)))?;
+        Ok(Self {
+            key_deregistration: KeyDeregistration::from_cbor_bytes(&dereg_metadatum.to_original_cbor_bytes())?,
+            deregistration_witness: DeregistrationWitness::from_cbor_bytes(&witness_metadatum.to_original_cbor_bytes())?,
+        })
+    }
+}
+
+impl std::convert::TryInto<Metadata> for &DeregistrationCbor {
+    type Error = DeserializeError;
+
+    fn try_into(self) -> Result<Metadata, Self::Error> {
+        let mut metadata = Metadata::new();
+        self.add_to_metadata(&mut metadata)?;
+        Ok(metadata)
+    }
+}
+
+
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub struct DeregistrationWitness {
+    pub stake_witness: StakeWitness,
+    #[serde(skip)]
+    pub encodings: Option<DeregistrationWitnessEncoding>,
+}
+
+impl DeregistrationWitness {
+    pub fn new(stake_witness: StakeWitness) -> Self {
+        Self {
+            stake_witness,
+            encodings: None,
+        }
+    }
+}
+
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub struct KeyDeregistration {
+    pub stake_credential: StakeCredential,
+    pub nonce: Nonce,
+    pub voting_purpose: VotingPurpose,
+    #[serde(skip)]
+    pub encodings: Option<KeyDeregistrationEncoding>,
+}
+
+impl KeyDeregistration {
+    /// Creates a new KeyDeregistration. You must then sign self.hash_to_sign() to make a `DeregistrationWitness`.
+    ///
+    /// # Arguments
+    ///
+    /// * `stake_credential` - stake address for the network that this transaction is submitted to (to point to the Ada that was being delegated).
+    /// * `nonce` - Monotonically rising across all transactions with the same staking key. Recommended to just use the slot of this tx.
+    pub fn new(stake_credential: StakeCredential, nonce: Nonce) -> Self {
+        Self {
+            stake_credential,
+            nonce,
+            voting_purpose: 0,
+            encodings: None,
+        }
+    }
+
+    /// Create bytes to sign to make a `DeregistrationWitness` from.
+    ///
+    /// # Arguments
+    ///
+    /// * `force_canonical` - Whether to encode the inner registration canonically. Should be true for hardware wallets and false otherwise.
+    pub fn hash_to_sign(&self, force_canonical: bool) -> cbor_event::Result<Vec<u8>> {
```

Comment:
Do we want to have a function that combines this all in one step with a key from `crypto` to directly produce the witness? I left this as-is in case people are using it as part of other libs or need to pass it elsewhere (is this ever needed? e.g. dApp connector or whatever or would this only ever be done in one step all in the wallet?)